### ISSUE TITLE
85 行 Python 代码实现极简版 GPT【from gemini2.5】

### ISSUE THREAD

**Issue by qianglifeng** [qianglifeng/IssueShared#1]
### 揭秘 GPT 工作原理：85 行 Python 代码实现极简版 GPT

**导言**

近年来，以 GPT（Generative Pre-trained Transformer）为代表的大型语言模型（LLM）取得了惊人的发展，能够生成媲美人类水平的文本、代码甚至艺术作品。然而，其内部复杂的运作机制常常令人望而生畏。

近期，一篇广受欢迎的博客文章以其精简的 Python 实现，为我们揭开了 GPT 神秘面纱的一角。这篇由 Jay Mody 撰写的文章，通过一个仅有 200 行代码的极简 GPT 模型，清晰地阐释了其核心工作流程。本文将对该实现进行编译和解读，帮助您深入理解 GPT 的内在逻辑。

值得一提的是，另一位人工智能领域的知名专家 Andrej Karpathy 也曾创建过一个名为 `minGPT` 的项目，并录制了详细的教学视频，旨在从零开始构建一个 GPT 模型。这些宝贵的学习资源都为我们理解 GPT 的工作原理提供了极大的便利。

-----

### **核心组件**

一个极简的 GPT 模型主要由以下几个核心组件构成，它们协同工作，从而实现文本的理解和生成：

1.  **词元化（Tokenization）：** 将输入的文本分解成模型能够理解的最小单元——词元（token）。
2.  **词嵌入（Token Embeddings）：** 将每个词元转换为一个高维度的向量表示。
3.  **位置编码（Positional Encodings）：** 向词嵌入中添加位置信息，使得模型能够理解单词在序列中的顺序。
4.  **Transformer 解码器模块（Transformer Decoder Blocks）：** 这是 GPT 的核心，通过自注意力机制（Self-Attention）和前馈神经网络（Feed-Forward Neural Network）来处理输入的序列。
5.  **语言模型头（Language Model Head）：** 将 Transformer 的输出转换为对下一个词元的预测概率分布。

-----

### **代码实现解析**

以下我们将结合一个精简至约 85 行的 Python 代码示例（基于 Jay Mody 和 Andrej Karpathy 的思想），来逐步解析 GPT 的工作流程。

```python
import torch
import torch.nn as nn
from torch.nn import functional as F

# --- 超参数 ---
batch_size = 64
block_size = 256
learning_rate = 3e-4
max_iters = 5000
eval_interval = 500
n_embd = 384
n_head = 6
n_layer = 6
dropout = 0.2
device = 'cuda' if torch.cuda.is_available() else 'cpu'

# --- 数据加载和预处理 ---
# 此处省略了数据加载的详细代码，通常包括读取文本、创建字符集、编码和解码函数等。
# 假设我们已经有了一个编码器（encoder）和解码器（decoder），以及训练和验证数据。

# --- 模型定义 ---
class Head(nn.Module):
    """一个自注意力头"""
    def __init__(self, head_size):
        super().__init__()
        self.key = nn.Linear(n_embd, head_size, bias=False)
        self.query = nn.Linear(n_embd, head_size, bias=False)
        self.value = nn.Linear(n_embd, head_size, bias=False)
        self.register_buffer('tril', torch.tril(torch.ones(block_size, block_size)))
        self.dropout = nn.Dropout(dropout)

    def forward(self, x):
        B, T, C = x.shape
        k = self.key(x)
        q = self.query(x)
        wei = q @ k.transpose(-2, -1) * C**-0.5
        wei = wei.masked_fill(self.tril[:T, :T] == 0, float('-inf'))
        wei = F.softmax(wei, dim=-1)
        wei = self.dropout(wei)
        v = self.value(x)
        out = wei @ v
        return out

class MultiHeadAttention(nn.Module):
    """多个自注意力头并行计算"""
    def __init__(self, num_heads, head_size):
        super().__init__()
        self.heads = nn.ModuleList([Head(head_size) for _ in range(num_heads)])
        self.proj = nn.Linear(n_embd, n_embd)
        self.dropout = nn.Dropout(dropout)

    def forward(self, x):
        out = torch.cat([h(x) for h in self.heads], dim=-1)
        out = self.dropout(self.proj(out))
        return out

class FeedFoward(nn.Module):
    """一个简单的前馈神经网络"""
    def __init__(self, n_embd):
        super().__init__()
        self.net = nn.Sequential(
            nn.Linear(n_embd, 4 * n_embd),
            nn.ReLU(),
            nn.Linear(4 * n_embd, n_embd),
            nn.Dropout(dropout),
        )

    def forward(self, x):
        return self.net(x)

class Block(nn.Module):
    """Transformer 模块：包含多头自注意力和前馈网络"""
    def __init__(self, n_embd, n_head):
        super().__init__()
        head_size = n_embd // n_head
        self.sa = MultiHeadAttention(n_head, head_size)
        self.ffwd = FeedFoward(n_embd)
        self.ln1 = nn.LayerNorm(n_embd)
        self.ln2 = nn.LayerNorm(n_embd)

    def forward(self, x):
        x = x + self.sa(self.ln1(x))
        x = x + self.ffwd(self.ln2(x))
        return x

class GPTLanguageModel(nn.Module):
    def __init__(self, vocab_size):
        super().__init__()
        self.token_embedding_table = nn.Embedding(vocab_size, n_embd)
        self.position_embedding_table = nn.Embedding(block_size, n_embd)
        self.blocks = nn.Sequential(*[Block(n_embd, n_head=n_head) for _ in range(n_layer)])
        self.ln_f = nn.LayerNorm(n_embd)
        self.lm_head = nn.Linear(n_embd, vocab_size)

    def forward(self, idx, targets=None):
        B, T = idx.shape
        tok_emb = self.token_embedding_table(idx)
        pos_emb = self.position_embedding_table(torch.arange(T, device=device))
        x = tok_emb + pos_emb
        x = self.blocks(x)
        x = self.ln_f(x)
        logits = self.lm_head(x)

        loss = None
        if targets is not None:
            B, T, C = logits.shape
            logits = logits.view(B*T, C)
            targets = targets.view(B*T)
            loss = F.cross_entropy(logits, targets)

        return logits, loss

    def generate(self, idx, max_new_tokens):
        for _ in range(max_new_tokens):
            idx_cond = idx[:, -block_size:]
            logits, loss = self(idx_cond)
            logits = logits[:, -1, :]
            probs = F.softmax(logits, dim=-1)
            idx_next = torch.multinomial(probs, num_samples=1)
            idx = torch.cat((idx, idx_next), dim=1)
        return idx
```

#### **1. 输入处理：词嵌入与位置编码**

当一串文本输入模型时，首先通过 `encoder` 将其转换为一串数字（词元 ID）。然后，`GPTLanguageModel` 的 `forward` 方法开始处理：

  - `self.token_embedding_table`：这是一个嵌入层，它将每个词元 ID 映射为一个高维向量。这使得模型能够以向量的形式理解和处理单词。
  - `self.position_embedding_table`：为了让模型理解单词的顺序，我们为每个位置（0 到 `block_size`-1）也创建一个嵌入向量。
  - `x = tok_emb + pos_emb`：将词元嵌入和位置嵌入相加，这样每个输入向量就同时包含了单词的语义信息和其在序列中的位置信息。

#### **2. Transformer 模块：信息处理的核心**

接下来，融合了位置信息的词向量 `x` 会被送入一系列的 `Block`（Transformer 解码器模块）中进行处理。每个 `Block` 包含两个关键的子层：

  - **多头自注意力（MultiHeadAttention）：**

      - **自注意力（Head）：** 这是 Transformer 的核心思想。对于序列中的每一个词元，自注意力机制会计算它与序列中所有其他词元（包括它自己）的关联程度，这个关联程度被称为“注意力权重”。
      - **Query, Key, Value：** 为了计算注意力权重，输入向量 `x` 会被线性变换成三个不同的向量：查询向量（Query）、键向量（Key）和值向量（Value）。可以通俗地理解为：Query 代表当前词元想要查询的信息，Key 代表序列中其他词元可以提供的信息标签，而 Value 则是这些词元实际包含的信息。通过计算 Query 和所有 Key 的相似度（点积），就可以得到注意力权重，然后用这个权重对所有的 Value 进行加权求和，从而得到一个融合了上下文信息的新的向量表示。
      - **掩码（Masking）：** 在生成文本时，模型在预测第 `T` 个词时，应该只能看到它前面的 `T-1` 个词，而不能看到未来的词。代码中的 `tril` 就是一个下三角矩阵，它用来实现这种“掩码”操作，防止模型“作弊”。
      - **多头（Multi-Head）：** 与其只进行一次自注意力计算，不如并行地进行多次（由 `n_head` 定义），每个“头”关注输入序列的不同方面。最后将所有头的结果拼接起来，从而获得更丰富的特征表示。

  - **前馈神经网络（FeedFoward）：**

      - 在自注意力层之后，每个位置的输出向量都会独立地通过一个简单的前馈神经网络。这可以看作是对自注意力机制获得的信息进行进一步的非线性加工和提炼。

  - **残差连接与层归一化（Residual Connections & Layer Normalization）：**

      - 在每个 `Block` 中，输入 `x` 会被直接加到子层（自注意力和前馈网络）的输出上（`x = x + self.sa(...)`），这被称为“残差连接”。这种设计有助于缓解深度神经网络中的梯度消失问题，使得模型可以堆叠更多的层。
      - `LayerNorm`（层归一化）则是在每个子层之前对输入进行归一化，使得训练过程更加稳定。

#### **3. 输出与生成：预测下一个词**

经过多层 `Block` 的处理后，我们得到了一个包含了丰富上下文信息的序列向量 `x`。

  - `self.lm_head`：这是一个线性层，它将 Transformer 模块的输出向量 `x` 投影到词汇表的大小。这样，对于序列中的每个位置，我们都得到了一个 logits 向量，表示词汇表中每个词元是下一个词的原始分数。
  - **损失计算：** 在训练过程中，我们将 logits 与真实的下一个词元（`targets`）进行比较，通过交叉熵损失（`F.cross_entropy`）来计算模型的预测与真实值之间的差距，并以此来更新模型的参数。
  - **文本生成：** 在 `generate` 方法中，模型会进行自回归式的生成。它首先根据输入的上下文预测下一个词的概率分布（通过对 logits 应用 `softmax` 函数），然后从这个分布中随机采样一个词元作为预测结果。接着，将这个新生成的词元拼接到输入序列的末尾，作为下一次预测的新上下文，如此循环，直到生成所需长度的文本。

### **结论**

这个极简的 GPT 实现虽然在规模和性能上无法与 GPT-3 或 GPT-4 等巨型模型相提并论，但它麻雀虽小，五脏俱全。它清晰地展示了 GPT 模型如何通过词嵌入、位置编码、多层 Transformer 模块（核心是自注意力机制）以及一个最终的线性层，来实现对文本的理解和生成。

通过学习和理解这样的代码实现，我们能够拨开大型语言模型复杂的外壳，洞察其背后优雅而强大的核心原理。
